### PR TITLE
Refactor PostTypeChecker into multiple classes per usecase

### PR DESCRIPTION
**Pull request by Marenz**

### CODE REVIEW

File: libsolidity/analysis/PostTypeChecker.cpp [hunk 1]
```diff
@@ -31,106 +31,179 @@ using namespace dev;
...
+	bool visit(ContractDefinition const&) override
+	{
+		solAssert(!m_currentConstVariable, "");
+		solAssert(m_constVariableDependencies.empty(), "");
+		return true;
+	}
+
+	void endVisit(ContractDefinition const&) override
+	{
+		solAssert(!m_currentConstVariable, "");
```
Comment (chriseth) [inline libsolidity/analysis/PostTypeChecker.cpp:87]: Maybe this is better readable if the implementations are below the class?

File: libsolidity/analysis/PostTypeChecker.cpp [hunk 2]
```diff
@@ -31,106 +31,179 @@ using namespace dev;
...
+		};
+		return CycleDetector<VariableDeclaration>(visitor).run(_startingFrom);
+	}
+
+private:
+	VariableDeclaration const* m_currentConstVariable = nullptr;
+	std::map<VariableDeclaration const*, std::set<VariableDeclaration const*>> m_constVariableDependencies;
+	std::vector<VariableDeclaration const*> m_constVariables; ///< Required for determinism.
+
+	ErrorReporter& m_errorReporter;
```
Comment (chriseth) [inline libsolidity/analysis/PostTypeChecker.cpp]: Could be moved into a common base class.

File: libsolidity/analysis/PostTypeChecker.cpp [hunk 3]
```diff
@@ -31,106 +31,179 @@ using namespace dev;
...
 }
 
 void PostTypeChecker::endVisit(VariableDeclaration const& _variable)
 {
-	if (_variable.isConstant())
-	{
-		solAssert(m_currentConstVariable == &_variable, "");
-		m_currentConstVariable = nullptr;
-	}
+	for (auto& checker: m_checkers)
```
Comment (chriseth) [inline libsolidity/analysis/PostTypeChecker.cpp]: Template function for the loop?

Comment (Marenz) [reply libsolidity/analysis/PostTypeChecker.cpp]: not sure what you mean?

Comment (Marenz) [reply libsolidity/analysis/PostTypeChecker.cpp]: ah. Understood

Comment (Marenz) [reply libsolidity/analysis/PostTypeChecker.cpp]: Ah, understood

**Comment by Marenz** [comment]
updated

File: libsolidity/analysis/PostTypeChecker.cpp [hunk 4]
```diff
@@ -31,106 +31,175 @@ using namespace dev;
 using namespace langutil;
 using namespace dev::solidity;
 
+// Defined at the bottom to avoid declaring all Checkers at the top
+vector<shared_ptr<PostTypeChecker::Checker>> initializeCheckers(ErrorReporter& _errorReporter);
```
Comment (chriseth) [inline libsolidity/analysis/PostTypeChecker.cpp]: Put inside anonymous namespace?

Comment (chriseth) [reply libsolidity/analysis/PostTypeChecker.cpp]: Another way would be to define the constructor further down...

File: libsolidity/analysis/PostTypeChecker.h [hunk 5]
```diff
@@ -61,13 +71,25 @@ class PostTypeChecker: private ASTConstVisitor
...
-	VariableDeclaration const* findCycle(VariableDeclaration const& _startingFrom);
+	template <class T>
+	void callEndVisit(T const& _node)
+	{
+		for (auto& checker: m_checkers)
+			checker->endVisit(_node);
+	}
+
+	template <class T>
+	bool callVisit(T const& _node)
```
Comment (chriseth) [inline libsolidity/analysis/PostTypeChecker.h]: Any reason you did not put `callVisit` above `callEndVisit`?

File: libsolidity/analysis/PostTypeChecker.cpp (hunk 4 above)
Comment (Marenz) [reply libsolidity/analysis/PostTypeChecker.cpp]: I didn't want to split the class member definitions into several blocks

**Comment by Marenz** [comment]
updated

**Review by erak — APPROVED** [review]

**Pull request merged**